vulnerabilities/phantom-functions.md

```diff
@@ -0,0 +1,134 @@
+
+        receiver.onFlashLoan(amount, data);
+
+        tokenX.burn(address(receiver), amount + fee);
+        
+        emit Flashloaned(receiver, amount, fee);
+    }
+}
+```
+
+The attack scenario goes like this:
+
+The victim is a contract called `VulnerableContract`, deployed via GnosisSafeProxy, which contains a fallback function that doesn't revert when calling a non-existent function. The user stores 100 tokenX in the `VulnerableContract` contract.
+
+1. The attacker does not need special permissions as the `flashloan` function is public.
+2. **Flashloan Function Call**: The attacker calls `flashloan` on `VulnerableContract` with `VulnerableContract` - itself, as the `receiver`.
+   - Example call:
+   
+```solidity
+vulnerableContractInstance.flashloan(VulnerableContract, 1000, "");
+```
+
+3. **Minting Tokens**: `VulnerableContract` mints 1000 `tokenX` to `VulnerableContract` - itself.
+4. **Calling `onFlashLoan`**: The `onFlashLoan` function is expected to be called on `VulnerableContract` - the receiver of the flashloan. Since `VulnerableContract` does not have this function implemented, its fallback function is invoked. The fallback function does not revert, allowing the transaction to proceed.
+5. **Burning Tokens**: `VulnerableContract` attempts to burn `1000 + 50 = 1050` `tokenX` from `VulnerableContract` - itself. `VulnerableContract` had 100 tokenX initially (plus the newly minted 1000 tokens). The burn function will reduce the total `tokenX` balance in `VulnerableContract` by 1050 tokens, effectively burning an extra 50 tokens that were not part of the amount!
+
+### Financial Loss
+
+The victim (VulnerableContract) suffers a loss as follows:
+
+1. **Initial Balance**: VulnerableContract starts with 100 `tokenX`.
+2. **After Minting**: VulnerableContract receives 1000 `tokenX`, resulting in a total of 1100 `tokenX`.
+3. **After Burning**: The contract attempts to burn 1050 `tokenX`. After burning, VulnerableContract is left with 50 tokens (1100 - 1050).
+4. **Net Loss**: The contract, VulnerableContract, loses 50 `tokenX` due to the fee imposed by the flashloan.
+
+## Mitigating Risks of Phantom Functions
+
+It is evident that having a fallback function inside a contract can introduce significant risks when integrating with external contracts. To mitigate these risks, developers must assume responsibility for user safety in all scenarios. The following are suggested ways to mitigate the dangers of phantom functions:
+
+- **Validate the Presence of Expected Functions**: Implement checks to ensure the `receiver` contract supports the expected functions. For instance, use interface checks to verify that the contract implements the required functions before calling them.
+- **Avoid Unsafe Integrations**: Refrain from integrating with contracts that have fallback functions unless absolutely necessary. If integration is required, thoroughly audit the external contract's code and behavior.
```

Comment:
I don't think we necessarily need to avoid integrating with contracts with fallback functions because as long as the function we intend to call is present, there's no risk

Reply:
That is why checking for the existence of the function the developer wants to invoke in the external contract before calling it as suggested in the "Mitigating Risks of Phantom Functions" is given as a mitigating strategy.

On avoiding Unsafe Integrations, the entirety of sentence I believe speaks for itself, i.e **Refrain from integrating with contracts that have fallback functions unless absolutely necessary. If integration is required, thoroughly audit the external contract's code and behavior**

Reply:
The thing is that the fallback functions existence isn't ever going to be the bug, but rather the lack of the intended function. Suggesting that we should avoid fallback functions doesn't actually improve the security since if execution simply reverted unexpectedly, that might be just as significant of a bug